File: 1615-range-sum-of-sorted-subarray-sums/range-sum-of-sorted-subarray-sums.cpp
class Solution {
public:
    int rangeSum(vector<int>& nums, int n, int left, int right) {
        const int MOD = 1e9 + 7
        ;int ans=0
        ;vector<int>l;
        for(int i =0;i<n;i++){
            int s=0
            ;for (int j =i;j<n;j++){
                s+=nums[j]
                ;l.push_back(s)
            ;}
        }
        sort(l.begin(),l.end())
        ;for(int i=left-1;i<right;i++){
            ans=(ans+l[i])%MOD
        ;}
        return ans
   ; }

};

File: 1615-range-sum-of-sorted-subarray-sums/range-sum-of-sorted-subarray-sums.py
class Solution:
    

    def rangeSum(self, nums: List[int], n: int, left: int, right: int) -> int:
        mod = 10**9 + 7

        # Generate all possible subarray sums
        subarray_sums = []
        for i in range(n):
            s = 0
            for j in range(i, n):
                s += nums[j]
                subarray_sums.append(s)

        # Find the sum of elements between 'left' and 'right' using heap
        # Use a min-heap to find the kth smallest element efficiently
        heapq.heapify(subarray_sums)
        result = 0

        for i in range(1, right + 1):
            val = heapq.heappop(subarray_sums)
            if i >= left:
                result += val
                result %= mod

        return result
